### PR TITLE
expression: fix out of range error for real number

### PR DESCRIPTION
**Pull request by mccxj**
for #4477

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/5295)

### CODE REVIEW

File: expression/builtin_arithmetic.go [hunk 1]
```diff
@@ -594,6 +603,7 @@ func (c *arithmeticIntDivideFunctionClass) getFunction(ctx context.Context, args
 }
 
 type builtinArithmeticIntDivideIntSig struct{ baseBuiltinFunc }
+type builtinArithmeticRealDivideRealSig struct{ baseBuiltinFunc }
```
Comment (coocood) [inline expression/builtin_arithmetic.go]: I think the `IntDivide` represents the 'DIV` operator, so the name should be `IntDivideRealSig`.

Comment (mccxj) [reply expression/builtin_arithmetic.go]: You are right. PTAL.

File: expression/builtin_arithmetic.go [hunk 2]
```diff
@@ -635,6 +645,26 @@ func (s *builtinArithmeticIntDivideIntSig) evalInt(row types.Row) (int64, bool,
...
 }
 
+func (s *builtinArithmeticRealDivideRealSig) evalInt(row types.Row) (int64, bool, error) {
+	sc := s.ctx.GetSessionVars().StmtCtx
+	b, isNull, err := s.args[1].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	if b == 0 {
```
Comment (zimulala) [inline expression/builtin_arithmetic.go]: You may need to handle the case as follows:
```
mysql> select (1 div 1E-82);
+---------------+
| (1 div 1E-82) |
+---------------+
|          NULL |
+---------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+---------------+
| Level   | Code | Message       |
+---------+------+---------------+
| Warning | 1365 | Division by 0 |
+---------+------+---------------+
1 row in set (0.00 sec)
```
Or you can add a `TODO` here.

File: expression/builtin_arithmetic.go (hunk 1 above)
Comment (zimulala) [inline expression/builtin_arithmetic.go]: s/builtinArithmeticRealDivideRealSig/builtinArithmeticIntDivideRealSig

**Comment by coocood** [comment]
LGTM

File: expression/builtin_arithmetic.go [hunk 3]
```diff
@@ -635,6 +645,26 @@ func (s *builtinArithmeticIntDivideIntSig) evalInt(row types.Row) (int64, bool,
 	return ret, err != nil, errors.Trace(err)
 }
 
+func (s *builtinArithmeticIntDivideRealSig) evalInt(row types.Row) (int64, bool, error) {
+	sc := s.ctx.GetSessionVars().StmtCtx
+	b, isNull, err := s.args[1].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: `return 0, true, errors.Trace(err)` is more specific

File: expression/builtin_arithmetic.go [hunk 4]
```diff
@@ -635,6 +645,26 @@ func (s *builtinArithmeticIntDivideIntSig) evalInt(row types.Row) (int64, bool,
...
+	}
+
+	if b == 0 {
+		return 0, true, errors.Trace(handleDivisionByZeroError(s.ctx))
+	}
+
+	a, isNull, err := s.args[0].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: ditto

File: types/overflow.go [hunk 5]
```diff
@@ -187,3 +187,13 @@ func DivIntWithUint(a int64, b uint64) (uint64, error) {
 
 	return uint64(a) / b, nil
 }
+
+// DivFloat64 divides float64 a with b, returns int64 if no overflow error.
+// It just checks overflow, if b is zero, a "divide by zero" panic throws.
+func DivFloat64(a float64, b float64) (int64, error) {
+	c := a / b
+	if c > math.MaxInt64 || c < math.MinInt64 {
+		return 0, ErrOverflow.GenByArgs("BIGINT", fmt.Sprintf("(%f, %f)", a, b))
```
Comment (zz-jason) [inline types/overflow.go]: add a test like "select (1 div 1E-20);" and check the error message.

**Comment by mccxj** [comment]
maybe I will revert `real` to the origin `decimal` case, and test the overflow case of `EvalDecimal`. WIP.

**Comment by zimulala** [comment]
@mccxj 
I add the ‘WIP’ tag to your PR. If your PR is ready to be reviewed, let us know.

**Comment by zz-jason** [comment]
@mccxj any update ?

**Comment by mccxj** [comment]
@zz-jason I will open another PR to continue working.

**Pull request closed without merging**